src/mds/MDSDmclockScheduler.cc

```diff
@@ -0,0 +1,891 @@
+  dout(10) << __func__ << " volume_id " << vid << " session_id " << sid << dendl;
+  delete_session_from_volume_info(vid, sid);
+}
+
+uint32_t MDSDmclockScheduler::get_request_queue_size() const
+{
+  std::unique_lock<std::mutex> lock(queue_mutex);
+  return request_queue.size();
+}
+
+void MDSDmclockScheduler::enqueue_update_request(const VolumeId& vid, const ClientInfo& client_info, const bool use_default)
+{
+    dout(10) << __func__ <<  " volume_id " << vid << dendl;
+
+    std::unique_lock<std::mutex> lock(queue_mutex);
+
+    request_queue.emplace_back(new UpdateRequest(vid, client_info, use_default));
+
+    /* wake up*/
+    lock.unlock();
+    queue_cvar.notify_all();
+}
+
+void MDSDmclockScheduler::enqueue_update_request(const VolumeId& vid, const ClientInfo& client_info, const bool use_default, RequestCB cb_func)
+{
+    dout(10) << __func__ <<  " volume_id " << vid << dendl;
+
+    std::unique_lock<std::mutex> lock(queue_mutex);
+
+    request_queue.emplace_back(new UpdateRequest(vid, client_info, use_default, cb_func));
+    /* wake up*/
+    lock.unlock();
+    queue_cvar.notify_all();
+}
+
+void MDSDmclockScheduler::process_request_handler()
+{
+  std::unique_lock<std::mutex> lock(queue_mutex);
+
+  if (request_queue.size() == 0) {
+    queue_cvar.wait(lock);
+  }
+
+  while (request_queue.size()) {
+    std::unique_ptr<Request> request = std::move(request_queue.front());
+    request_queue.erase(request_queue.begin());
+
+    dout(10) << __func__ << " Process request type " << request->get_request_type_str() << dendl;
+
+    lock.unlock();
+
+    switch(request->get_request_type()) {
+      case RequestType::CLIENT_REQUEST:
+      {
+        std::unique_ptr<ClientRequest> c_request(static_cast<ClientRequest *>(request.release()));
+
+        dout(10) << " Process client request (queue size " << request_queue.size()
+                << " volume_id " << c_request->get_volume_id()
+                << " time " << c_request->time
+                << " cost " << c_request->cost << ")" << dendl;
+
+        increase_inflight_request(c_request->get_volume_id());
+
+        dmclock_queue->add_request(std::move(c_request), std::move(c_request->get_volume_id()),
+            {0, 0}, c_request->time, c_request->cost);
+
+        break;
+      }
+      case RequestType::UPDATE_REQUEST:
+      {
+        std::unique_ptr<UpdateRequest> c_request(static_cast<UpdateRequest *>(request.release()));
+
+        dout(10) << " Process update request (queue size " << request_queue.size()
+                  << " volume_id " << c_request->get_volume_id() << ")" << dendl;
+
+        update_volume_info(c_request->get_volume_id(), c_request->client_info, c_request->use_default);
+        dmclock_queue->update_client_info(c_request->get_volume_id());
+        set_volume_info_updated(c_request->get_volume_id());
+
+        if (c_request->cb_func) {
+          c_request->cb_func();
+        }
+        break;
+      }
+    }
```

Comment:
personal nit: case statement should have a `default` block that processes or throws an error

Reply:
It seems that there are only two types, RequestType::CLIENT_REQUEST and RequestType::UPDATE_REQUEST, and there are no exceptional situations expected. However, to be prepared, I will add a default case.

https://github.com/yongseokoh/ceph/blob/mds-dmclock-scheduler/src/mds/MDSDmclockScheduler.h#L101
https://github.com/yongseokoh/ceph/blob/mds-dmclock-scheduler/src/mds/MDSDmclockScheduler.h#L112